### ISSUE TITLE
DNN trainer error on varying batch sizes during optimization

### ISSUE THREAD

**Issue by kkornel1977** [davisking/dlib#1409]
Hello! First of all, thanks for this library. It is really useful and makes the everyday developments much easier for me!

## Expected Behavior
I use dlib/dnn package to process vector sequences of different length. During optimization the network parameters could be updated after processing a sequence: evaluate the network on sequence, compute gradient and finally backpropagate error with my own loss function. In dlib terminology, one sequence will be a mini-batch, and therefore, every mini-batch could have different size than the previous one. Furtunately, the dnn_trainer class train_one_step() member function makes the implementation easy, but it fails in several cases.

## Current Behavior
First call of train_one_step() is always executed as expected. If next call has different mini-batch size, then the call fails depending on the following circumstances: i) type of the first layer on input, ii) numbers of gpu devices.

## Steps to Reproduce
Consider the following short C++ sample code. It generates a regression task with random 28x28 images and labels. The network applies an arbitrary layer on the input, then a simple full connected regression output. Network optimization is done through several epoch with (varying size) mini-batches and Adam optimizer.

```
#include <dlib/dnn.h>
#include <vector>

using namespace dlib;

#define MINI_BATCH_SIZE_MIN 128
#define MINI_BATCH_SIZE_MAX 1024

// works as first layer: con<1,1,1,1,1, tag1<
// throws error or core dumped as first layer: affine< bn< dropout< htan< relu< sig< etc.
#define FIRST_LAYER relu<

using net_type = loss_mean_squared<fc<1,fc<256,
        FIRST_LAYER
        input<matrix<double>>>>>>;

#define GPU_DEVICE_LIST {0}
 
// testing non-constant batch size training
int main(int argc, char** argv)
{
	int epoch;
    dlib::rand rnd(time(0));
	
    // create a random "corpus"
    std::vector<matrix<double>> samples;
    std::vector<float> labels;
    
    while (samples.size()<MINI_BATCH_SIZE_MAX)
    {
    	samples.push_back(randm(28,28,rnd));
    	labels.push_back(rnd.get_random_float());
    }
    
    // create a net and prepare trainer
    net_type net;
    dnn_trainer<net_type,adam> trainer(net,adam(0.0005, 0.9, 0.999),GPU_DEVICE_LIST);

    trainer.set_learning_rate(1e-4);
    trainer.set_learning_rate_shrink_factor(1.0);
    trainer.be_quiet();
    
    // train
    for(epoch=0; epoch<25; epoch++)
    {
    	// update network with a random sized mini-batch
    	int mini_batch_size=MINI_BATCH_SIZE_MIN+(MINI_BATCH_SIZE_MAX<=MINI_BATCH_SIZE_MIN ? 0 : rnd.get_random_32bit_number()%(MINI_BATCH_SIZE_MAX-MINI_BATCH_SIZE_MIN));
    	trainer.train_one_step(samples.begin(), samples.begin()+mini_batch_size, labels.begin());
    	
    	// show something...
		std::cout << "epoch " << epoch+1 << ", loss: " << trainer.get_average_loss() << std::endl;
		trainer.clear_average_loss();
	}
	trainer.get_net();
		
	// done
	return 0;
}
```

In the example directory of dlib, it could be easily compiled without any extra dependencies, for example:
```
g++ -std=c++11 -O3 -I.. ../dlib/all/source.cpp -lpthread -lX11 main.cpp
```

When mini-batch size is constant in every epoch (i.e. `MINI_BATCH_SIZE_MIN` and `MINI_BATCH_SIZE_MAX` is equal), any layer could be used as `FIRST_LAYER` on input. But, when mini-batch size is varying epoch to epoch, only convolution and tagged layer is working in recent dlib versions, from version 19.11.
Dlib versions until 19.10 are working with any layer type on one GPU, but if we use multi-gpu environment (i.e. `GPU_DEVICE_LIST` contains more device ids) there is no version and layer, which could run without error.

* **Version**: 19.4 - 19.14
* **Where did you get dlib**: github repo
* **Platform**: CentOS 7; CUDA 9.1.85; CuDNN 7.1.2.21
* **Compiler**: gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28)

**Comment by davisking** [comment]
Oops, that's not right :)

Thanks for reporting this.  I just pushed the fix so it should all be working again.

**Comment by kkornel1977** [comment]
That's excellent! Thank you for the incredible quick fix!
